fs-index/src/watch.rs

```diff
@@ -0,0 +1,132 @@
+use std::{fs, path::Path};
+
+use anyhow::Result;
+use futures::{
+    channel::mpsc::{channel, Receiver},
+    SinkExt, StreamExt,
+};
+use log::info;
+use notify::{Config, Event, RecommendedWatcher, RecursiveMode, Watcher};
+
+use data_resource::ResourceId;
+use fs_storage::ARK_FOLDER;
+
+use crate::ResourceIndex;
+
+/// Watches a given directory for file system changes and automatically updates
+/// the resource index.
+///
+/// This function continuously monitors the specified directory and responds to
+/// file system events such as file creation, modification, and deletion. When
+/// an event is detected, the function updates the associated resource index and
+/// stores the changes.
+///
+/// The function runs asynchronously, whcih makes it suitable for non-blocking
+/// contexts. It uses a recursive watcher to track all changes within the
+/// directory tree. Events related to the internal `.ark` folder are ignored to
+/// prevent unnecessary updates.
+///
+/// # Arguments
+///
+/// * `root_path` - The root directory to be watched. This path is canonicalized
+///   to handle symbolic links and relative paths correctly.
+pub async fn watch_index<P: AsRef<Path>, Id: ResourceId>(
+    root_path: P,
+) -> Result<()> {
```

Comment:
This public function seems is the only API entrypoint, but in fact developer cannot extract any information from running the watcher. How would a developer integrate with Watch API?

I think, we need some kind of stream or events emitter so the developer would be able to handle the indexing events. The content of the stream should be same as output of `update_all`.

Reply:
Updated `watch_index()` to return a stream of `WatchEvent<Id>`, where `WatchEvent` is:
```rust
/// Represents the different kinds of events that can occur when watching the
/// resource index.
#[derive(Debug)]
pub enum WatchEvent<Id: ResourceId> {
    /// Represents an update to a single resource.
    UpdatedOne(IndexUpdate<Id>),
    /// Represents an update to all resources.
    UpdatedAll(IndexUpdate<Id>),
}
```
This contains information about changes, whether it involves a single updated resource (`UpdatedOne`) or all resources (`UpdatedAll`).